core/src/main/java/org/infinispan/marshall/core/AbstractBytesObjectInput.java

```diff
@@ -0,0 +1,263 @@
+   @Override
+   public String readUTF() throws EOFException {
+      int utflen = readInt();
+
+      checkPosLength();
+      byte[] bytearr = bytes;
+      char[] chararr = new char[utflen];
+
+      utflen = pos + utflen;
+      int c, char2, char3;
+      int count = pos;
+      int chararr_count=0;
+
+      while (count < utflen) {
+         c = (int) bytearr[count] & 0xff;
+         if (c > 127) break;
+         count++;
+         chararr[chararr_count++]=(char)c;
+      }
+
+      while (count < utflen) {
+         c = (int) bytearr[count] & 0xff;
+         switch (c >> 4) {
+            case 0: case 1: case 2: case 3: case 4: case 5: case 6: case 7:
+               /* 0xxxxxxx*/
+               count++;
+               chararr[chararr_count++]=(char)c;
+               break;
+            case 12: case 13:
+               /* 110x xxxx   10xx xxxx*/
+               count += 2;
+               if (count > utflen)
+                  throw new RuntimeException(
+                        "malformed input: partial character at end");
+               char2 = (int) bytearr[count-1];
+               if ((char2 & 0xC0) != 0x80)
+                  throw new RuntimeException(
+                        "malformed input around byte " + count);
+               chararr[chararr_count++]=(char)(((c & 0x1F) << 6) |
+                     (char2 & 0x3F));
+               break;
+            case 14:
+               /* 1110 xxxx  10xx xxxx  10xx xxxx */
+               count += 3;
+               if (count > utflen)
+                  throw new RuntimeException(
+                        "malformed input: partial character at end");
+               char2 = (int) bytearr[count-2];
+               char3 = (int) bytearr[count-1];
+               if (((char2 & 0xC0) != 0x80) || ((char3 & 0xC0) != 0x80))
+                  throw new RuntimeException(
+                        "malformed input around byte " + (count-1));
+               chararr[chararr_count++]=(char)(((c     & 0x0F) << 12) |
+                     ((char2 & 0x3F) << 6)  |
+                     (char3 & 0x3F));
+               break;
+            default:
+               /* 10xx xxxx,  1111 xxxx */
+               throw new RuntimeException(
+                     "malformed input around byte " + count);
+         }
+      }
+
+      pos = count;
+
+      return new String(chararr, 0, chararr_count);
+   }
+
+   public String readString() throws EOFException {
+      int mark = readByte();
+
+      switch(mark) {
+         case 0:
+            return ""; // empty string
+         case 1:
+            // small ascii
+            int size = readByte();
+            String str = new String(bytes, 0, pos, size);
+            pos += size;
+            return str;
+         case 2:
+            // large string
+            return readUTF();
+         default:
+            throw new RuntimeException("Unknown marker(String). mark=" + mark);
+      }
+   }
+
+   private void checkPosLength() throws EOFException {
```

Comment:
This isn't quite correct, we should pass in an int argument that we add to pos to make sure we don't overflow. This only works for a single byte read.